0023-communication-protocol.adoc

```diff
@@ -0,0 +1,388 @@
+This imposes a lot of complexity on the message format, as it has to handle all kinds of different scenarios.
+Ideally, the new communication protocol has a more modular approach to allow for different messages to be sent, depending on the ledgers, assets and swap-protocols being involved.
+In a way, this relates to the open-closed principle.
+Extensions to the communication protocol should happen in the form of new messages instead of modifying existing ones.
+
+=== Support for more ledgers
+
+While designing the new protocol, one should take into account the needs for support Lightning and Grin.
+
+== Research
+
+=== Assumptions and goals
+
+I've started the research and design of the new communication protocol with the following assumptions and goals in mind.
+
+. Avoid interaction as part of the communication protocol.
+This makes the implementation easier and more resilient as we don't have to maintain in-memory state for each connection.
+. Favor small, single-purpose protocols over more complex ones.
+. In every context, it is clear which party will initiate the communication.
+This applies to opening the first substream that announces the swap and not to the actual network communication.footnote:[In libp2p, a single network connection is reused between two parties for several substreams.]
+We assume that the two parties have already established the roles of (cryptographic) Alice and Bob and those roles are an input to the communication protocol.
+
+To design a better communication protocol than the first one, it is useful to take a step back and think protocols in a more abstract manner.
+
+=== Protocols in COMIT
+
+A protocol can be roughly defined as the order and semantics of actions.
+
+Within COMIT, we have two kinds of protocols:
+
+. Cryptographic protocols
+. Communication protocols
+
+A cryptographic protocol within COMIT defines the order and semantics of interactions with blockchains.
+At the moment, COMIT specializes in cryptographic protocols that facilitate atomic swaps.
+
+With the https://github.com/comit-network/spikes/blob/master/0021-protocol-naming.adoc[protocol naming spike], we've made it more explicit that COMIT does not support one but several cryptographic protocols at the moment.
+In particular, we currently support:
+
+* HAN for Bitcoin
+* HAN for Ethereum
+* HErc20 for Ethereum
+
+We've managed to identify a protocol (HAN) that can be applied to both of the ledgers that we currently support.
+Even though this allows for some reuse, when you take a look at the actual interactions with the blockchain, HAN for Bitcoin and HAN for Ethereum differ (you cannot broadcast a Bitcoin transaction on Ethereum and vice versa).
+Given that there is always two protocols that need to be combined for an atomic swap, our total number of supported protocols (considering the full swap as a black box) is 6 (2 * 3).footnote:[We don't allow HAN(Ethereum)-HErc20 at the moment, so this number drops to 4 in reality.]
```

Comment:
I don't understand the math here?  Is this the combinations of protocols that we support?  If so shouldn't it by 4, since we only support cross-chain swaps?

Reply:
Oh, this may be a typo.  You say 4 below.

Reply:
For clarification:

He means these 4:

```
HAN(Bitcoin)-HAN(Ether)
HAN(Ether)-HAN(Bitcoin)
HErc20-HAN(Bitcoin)
HAN(Bitcoin)-HErc20
```

+2 theoretical ones that do not make sense because you can just do that on Ethereum without needing atomic swaps:

```
HErc20-HAN(Ether)
HAN(Ether)-HErc20
```